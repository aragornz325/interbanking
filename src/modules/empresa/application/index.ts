import { CrearEmpresaUseCase } from './use-cases/crear-empresa.use-case';
import { ListarEmpresasAdheridasUltimoMesUseCase } from './use-cases/listar-adhesiones.use-case';
import { ListarEmpresasConTransferenciasUltimoMesUseCase } from './use-cases/listar-transferencias.use-case';

export const useCases = [
  CrearEmpresaUseCase,
  ListarEmpresasConTransferenciasUltimoMesUseCase,
  ListarEmpresasAdheridasUltimoMesUseCase,
];
