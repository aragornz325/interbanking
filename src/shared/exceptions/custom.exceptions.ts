import { BadRequestException, ConflictException } from '@nestjs/common';
import { HttpException, HttpStatus } from '@nestjs/common';

/**
 * Excepci칩n: DuplicateResourceException
 *
 * 丘멆잺 Se lanza cuando se intenta crear un recurso que ya existe.
 * Representa un error 409 - Conflict.
 *
 * @example
 * throw new DuplicateResourceException('El CUIT ya est치 registrado');
 */
export class DuplicateResourceException extends ConflictException {
  constructor(message = 'El recurso ya existe') {
    super(message);
  }
}

/**
 * Excepci칩n: InvalidForeignKeyException
 *
 * 游뛂 Se lanza cuando se intenta asociar una entidad con una clave for치nea inv치lida.
 * Representa un error 400 - Bad Request.
 *
 * @example
 * throw new InvalidForeignKeyException('La empresa no existe');
 */
export class InvalidForeignKeyException extends BadRequestException {
  constructor(message = 'Referencia inv치lida') {
    super(message);
  }
}

export class TooManyRequestsException extends HttpException {
  constructor(message = 'Too Many Requests') {
    super(message, HttpStatus.TOO_MANY_REQUESTS);
  }
}
