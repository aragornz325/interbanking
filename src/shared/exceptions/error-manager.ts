import {
  BadRequestException,
  InternalServerErrorException,
} from '@nestjs/common';
import { QueryFailedError } from 'typeorm';

import {
  DuplicateResourceException,
  InvalidForeignKeyException,
} from './custom.exceptions';

interface PostgresError {
  driverError: {
    code: string;
  };
}

export class ErrorManager {
  static handle(error: unknown): never {
    const isPostgresError =
      error instanceof QueryFailedError ||
      (typeof error === 'object' &&
        error !== null &&
        'driverError' in error &&
        typeof (error as any).driverError?.code === 'string');

    if (isPostgresError) {
      const code = (error as PostgresError).driverError.code;

      switch (code) {
        case '23505':
          throw new DuplicateResourceException();

        case '23503':
          throw new InvalidForeignKeyException();

        case '23502':
          throw new BadRequestException('Campo obligatorio no puede ser nulo');

        default:
          throw new InternalServerErrorException(
            'Error de base de datos desconocido',
          );
      }
    }

    // âœ… Relanzamos si no se pudo manejar, asegurando tipo
    throw error instanceof Error ? error : new Error('Unknown error');
  }
}
