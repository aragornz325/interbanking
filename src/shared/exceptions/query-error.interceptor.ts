import {
  Injectable,
  NestInterceptor,
  ExecutionContext,
  CallHandler,
  ConflictException,
  BadRequestException,
  InternalServerErrorException,
} from '@nestjs/common';
import { Observable, throwError } from 'rxjs';
import { catchError } from 'rxjs/operators';

@Injectable()
export class QueryErrorInterceptor implements NestInterceptor {
  intercept(context: ExecutionContext, next: CallHandler): Observable<any> {
    return next.handle().pipe(
      catchError((err) => {
        // Detectamos por estructura si es un error de TypeORM / Postgres
        const code = err?.driverError?.code;

        if (code) {
          switch (code) {
            case '23505':
              console.log('[INTERCEPTOR] Clave duplicada interceptada');
              return throwError(
                () =>
                  new ConflictException(
                    'El recurso ya existe (violación de clave única)',
                  ),
              );

            case '23503':
              return throwError(
                () =>
                  new BadRequestException(
                    'Referencia inválida (clave foránea)',
                  ),
              );

            case '23502':
              return throwError(
                () =>
                  new BadRequestException(
                    'Campo obligatorio no puede ser nulo',
                  ),
              );

            default:
              return throwError(
                () =>
                  new InternalServerErrorException('Error de base de datos'),
              );
          }
        }

        return throwError(() => err);
      }),
    );
  }
}
