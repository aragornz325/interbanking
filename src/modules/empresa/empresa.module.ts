import { Module } from '@nestjs/common';
import { TypeOrmModule } from '@nestjs/typeorm';
// Token del repositorio
import { EMPRESA_REPOSITORY } from 'src/modules/empresa/domain/empresa.repository';

// Casos de uso (application layer)
import { CrearEmpresaUseCase } from './application/use-cases/crear-empresa.use-case';
import { ListarEmpresasAdheridasUltimoMesUseCase } from './application/use-cases/listar-adhesiones.use-case';
import { ListarEmpresasConTransferenciasUltimoMesUseCase } from './application/use-cases/listar-transferencias.use-case';
// Infraestructura y persistencia
import {
  EmpresaOrmEntity,
  EmpresaRepositoryImpl,
  TransferenciaOrmEntity,
} from './infrastructure/persistence';
// Presentaci√≥n
import { EmpresaController } from './presentation/controllers/empresa.controller';

@Module({
  imports: [
    TypeOrmModule.forFeature([EmpresaOrmEntity, TransferenciaOrmEntity]),
  ],
  controllers: [EmpresaController],
  providers: [
    {
      provide: EMPRESA_REPOSITORY,
      useClass: EmpresaRepositoryImpl,
    },
    CrearEmpresaUseCase,
    ListarEmpresasAdheridasUltimoMesUseCase,
    ListarEmpresasConTransferenciasUltimoMesUseCase,
  ],
  exports: [],
})
export class EmpresaModule {}
