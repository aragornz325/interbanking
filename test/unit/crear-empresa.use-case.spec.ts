// Jest setup file to configure the testing environment

// Nest core
import { Logger } from '@nestjs/common';

// Domain / Application
import { CrearEmpresaUseCase } from '../../src/modules/empresa/application/use-cases/crear-empresa.use-case';
import { Empresa } from '../../src/modules/empresa/domain/empresa.entity';
import { DuplicateResourceException } from 'src/shared/exceptions/custom.exceptions';

describe('CrearEmpresaUseCase', () => {
  let useCase: CrearEmpresaUseCase;
  let repo: { crear: jest.Mock };

  const cuit = '20304050607';
  const razonSocial = 'UNSC Spartan Ops';
  const fecha = new Date('2024-01-01');

  const empresaData = Empresa.createWithoutId(cuit, razonSocial, fecha);

  const empresaResult = new Empresa('uuid-1234', cuit, razonSocial, fecha);

  beforeEach(() => {
    repo = {
      crear: jest.fn().mockResolvedValue(empresaResult),
    };
    useCase = new CrearEmpresaUseCase(repo as any);
  });

  it('debe crear una empresa correctamente', async () => {
    const logSpy = jest
      .spyOn(Logger.prototype, 'log')
      .mockImplementation(() => {});

    const result = await useCase.execute(cuit, razonSocial);

    expect(result).toEqual(empresaResult);
    expect(repo.crear).toHaveBeenCalled();
    expect(logSpy).toHaveBeenCalledWith(
      expect.stringContaining('crearEmpresa'),
    );

    logSpy.mockRestore();
  });

  it('debe manejar errores lanzados por el repositorio', async () => {
    const mockError = new Error('DB fail');
    repo.crear.mockRejectedValueOnce(mockError);

    await expect(useCase.execute(cuit, razonSocial)).rejects.toThrow('DB fail');
  });

  it('debe rechazar CUITs duplicados (mockeando lÃ³gica)', async () => {
    const duplicadoError = new DuplicateResourceException(
      'Empresa ya registrada',
    );

    repo.crear.mockRejectedValueOnce(duplicadoError);

    try {
      await useCase.execute(cuit, razonSocial);
    } catch (err) {
      expect(err).toBeInstanceOf(DuplicateResourceException);
      expect((err as Error).message).toBe('Empresa ya registrada');
    }
  });

  it('debe hacer snapshot del resultado', async () => {
    const result = await useCase.execute(cuit, razonSocial);
    expect(result).toMatchSnapshot();
  });
});
