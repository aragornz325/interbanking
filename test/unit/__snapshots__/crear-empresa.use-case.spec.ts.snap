// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CrearEmpresaUseCase debe hacer snapshot del resultado 1`] = `
Empresa {
  "cuit": "20304050607",
  "fechaAdhesion": 2024-01-01T00:00:00.000Z,
  "id": "uuid-1234",
  "razonSocial": "UNSC Spartan Ops",
}
`;
